src/rust/engine/graph/src/lib.rs

```diff
@@ -514,17 +541,21 @@ impl<N: Node> Graph<N> {
           // TODO: doing cycle detection under the lock... unfortunate, but probably unavoidable
           // without a much more complicated algorithm.
```

Comment:
Unrelated: I would be interested in any thoughts on how to estimate the speedup we might get from incremental cycle detection (possibly just by using a profiler?) instead of holding the lock.